src/Parsers/FunctionSecretArgumentsFinderAST.h

```diff
@@ -1,549 +1,111 @@
-                }
-            }
-        }
 
-        /// We replace 'aws_secret_access_key' with '[HIDDEN]' for the following signatures:
-        /// S3('url', 'aws_access_key_id', 'aws_secret_access_key')
-        /// S3('url', 'aws_access_key_id', 'aws_secret_access_key', 'format')
-        /// S3('url', 'aws_access_key_id', 'aws_secret_access_key', 'format', 'compression')
-        if (2 < count)
-            markSecretArgument(2);
-    }
+        const auto * expr_list = function->arguments->as<ASTExpressionList>();
+        if (!expr_list)
+            return;
 
-    void findDatabaseEngineSecretArguments()
-    {
-        const String & engine_name = function.name;
-        if ((engine_name == "MySQL") || (engine_name == "MaterializeMySQL") ||
-            (engine_name == "MaterializedMySQL") || (engine_name == "PostgreSQL") ||
-            (engine_name == "MaterializedPostgreSQL"))
-        {
-            /// MySQL('host:port', 'database', 'user', 'password')
-            /// PostgreSQL('host:port', 'database', 'user', 'password')
-            findMySQLDatabaseSecretArguments();
-        }
-        else if (engine_name == "S3")
-        {
-            /// S3('url', 'access_key_id', 'secret_access_key')
-            findS3DatabaseSecretArguments();
-        }
+        arguments = std::make_unique<ArgumentsAST>(&expr_list->children);
     }
 
-    void findMySQLDatabaseSecretArguments()
-    {
-        if (isNamedCollectionName(0))
-        {
-            /// MySQL(named_collection, ..., password = 'password', ...)
-            findSecretNamedArgument("password", 1);
-        }
-        else
-        {
-            /// MySQL('host:port', 'database', 'user', 'password')
-            markSecretArgument(3);
-        }
-    }
+    String name() const override { return function->name; }
+private:
+    const ASTFunction * function = nullptr;
+};
 
-    void findS3DatabaseSecretArguments()
+/// Finds arguments of a specified function which should not be displayed for most users for security reasons.
+/// That involves passwords and secret keys.
+class FunctionSecretArgumentsFinderAST : public FunctionSecretArgumentsFinder
+{
+public:
+    explicit FunctionSecretArgumentsFinderAST(const ASTFunction & function_)
+        : FunctionSecretArgumentsFinder(std::make_unique<FunctionAST>(function_))
     {
-        if (isNamedCollectionName(0))
-        {
-            /// S3(named_collection, ..., secret_access_key = 'password', ...)
-            findSecretNamedArgument("secret_access_key", 1);
-        }
-        else
-        {
-            /// S3('url', 'access_key_id', 'secret_access_key')
-            markSecretArgument(2);
-        }
-    }
+        if (!function->hasArguments())
+            return;
 
-    void findBackupNameSecretArguments()
-    {
-        const String & engine_name = function.name;
-        if (engine_name == "S3")
+        switch (function_.kind)
```

Comment:
nit

```suggestion
        switch (function.kind)
```

Reply:
no, `function` is already abstracted object, and I did not populate `kind` for it - we don't need it anywhere else

Reply:
oh, right...